File: DownloadLib/src/main/java/com/kronos/download/RxDownload.java
package com.kronos.download;

import com.kronos.download.adapter.IObserver;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;

public class RxDownload {

    public static Observable<Integer> getDownload(String url) {
        DownloadModel downloadModel = DownloadManager.getInstance().getModel(url);
        RxObserver rxObserver = new RxObserver(downloadModel);
        Observable<Integer> observable = Observable.create(rxObserver);
        downloadModel.registerDataSetObserver(rxObserver);
        return observable;
    }


    private static class RxObserver implements IObserver, ObservableOnSubscribe<Integer> {
        private ObservableEmitter<Integer> emitter;
        DownloadModel model;

        RxObserver(DownloadModel model) {
            this.model = model;
        }

        @Override
        public void onChanged() {
            emitter.onNext(model.getProgress());
            if (model.check()) {
                emitter.onComplete();
            }
        }

        @Override
        public void subscribe(ObservableEmitter<Integer> emitter) throws Exception {
            this.emitter = emitter;
        }
    }
}

File: DownloadLib/src/main/java/com/kronos/download/RxDownload.kt
package com.kronos.download

import com.kronos.download.adapter.IObserver

import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import io.reactivex.ObservableOnSubscribe

object RxDownload {

    fun getDownload(url: String): Observable<Int> {
        val downloadModel = DownloadManager.getModel(url)
        val rxObserver = RxObserver(downloadModel)
        val observable = Observable.create(rxObserver)
        downloadModel!!.registerDataSetObserver(rxObserver)
        return observable
    }


    private class RxObserver internal constructor(internal var model: DownloadModel?) : IObserver, ObservableOnSubscribe<Int> {
        private var emitter: ObservableEmitter<Int>? = null

        override fun onChanged() {
            emitter!!.onNext(model?.progress!!)
            if (model!!.check()) {
                emitter!!.onComplete()
            }
        }

        @Throws(Exception::class)
        override fun subscribe(emitter: ObservableEmitter<Int>) {
            this.emitter = emitter
        }
    }
}
